### ISSUE TITLE
Integrate Sigma rules via OpenSearch Security Analytics

### ISSUE THREAD

**Issue by mmguero** [cisagov/Malcolm#475]
tasks:

* [ ] ❗❗❗resolve problems outlined in [this thread](https://forum.opensearch.org/t/trouble-with-mappings-detectors-and-alerts/23373) and [these bugs](https://github.com/opensearch-project/security-analytics/issues?q=is%3Aissue%20author%3Ammguero)
* [x] solve or work around issue with `nested` field types (see [comment below](https://github.com/cisagov/Malcolm/issues/475#issuecomment-2637723880))
* [ ] figure out mapping schema (probably just examine Sigma-used fields where we're not adhering perfectly to ECS and use the ECS mapping schema)
  * [ ] or create a [custom mapping](https://opensearch.org/docs/latest/security-analytics/api-tools/mappings-api/#create-mappings)
* [x] figure out aliases (as that's what [Security Analytics supports](https://opensearch.org/docs/latest/security-analytics/sec-analytics-config/detectors-config/#step-1-define-a-detector)) and automatic management of them (for both Opensearch and Elasticsearch) (done via mmguero-dev/Malcolm@ba0e38f2c01847a177b08721cd2ac61a0fbfca2c)

**[<img src="https://avatars.githubusercontent.com/u/13872653?s=17&v=4" width="17" height="17"> @mmguero](https://github.com/mmguero)** cloned issue [idaholab/Malcolm#162](https://github.com/idaholab/Malcolm/issues/162) on 2023-03-14: 

> How could Malcolm integrate sigma?
> 
> 
> 
> [Sigma](https://github.com/SigmaHQ/sigma):
> 
> 
> 
> > Sigma is a generic and open signature format that allows you to describe relevant log events in a straightforward manner. The rule format is very flexible, easy to write and applicable to any type of log file. The main purpose of this project is to provide a structured form in which researchers or analysts can describe their once developed detection methods and make them shareable with others.
> 
> > 
> 
> > Sigma is for log files what Snort is for network traffic and YARA is for files.
> 
> 
> 
> * [OpenSearch Security Analytics](https://opensearch.org/docs/2.4/security-analytics/index/) plugin uses Sigma. Malcolm actually includes this plugin just because it's part of OpenSearch, but we're not really doing anything with it.
> 
> * I think Elastic SIEM does as well (?)

**Comment by mmguero** [comment]
**[<img src="https://avatars.githubusercontent.com/u/53797?s=17&v=4" width="17" height="17"> @mavam](https://github.com/mavam)** [commented](https://github.com/idaholab/Malcolm/issues/162#issuecomment-1468693878) on 2023-03-14: 

> The current sigmac is going to be deprecated by the end of the year. If you start with this now, I would highly recommend going the pySigma route. That said, I don't know whether the [Elasticsearch backend](https://github.com/SigmaHQ/pySigma-backend-elasticsearch) also supports OpenSearch.
> 
> 
> 
> You could also consider writing a transpiler, as we did in [VAST](https://vast.io/docs/understand/language/frontends/sigma).

**Comment by mmguero** [comment]
**[<img src="https://avatars.githubusercontent.com/u/13872653?s=17&v=4" width="17" height="17"> @mmguero](https://github.com/mmguero)** [commented](https://github.com/idaholab/Malcolm/issues/162#issuecomment-1468707967) on 2023-03-14: 

> Thanks for the suggestion!

**Comment by mmguero** [comment]
**[<img src="https://avatars.githubusercontent.com/u/13872653?s=17&v=4" width="17" height="17"> @mmguero](https://github.com/mmguero)** [commented](https://github.com/idaholab/Malcolm/issues/162#issuecomment-2391791233) on 2024-10-03: 

> * [Sigma taxonomy](https://github.com/SigmaHQ/sigma-specification/blob/main/appendix/sigma-taxonomy-appendix.md)
>     * for mapping, note how [product](https://github.com/SigmaHQ/sigma-specification/blob/main/appendix/sigma-taxonomy-appendix.md#zeek) and service seem pretty much to line up with our `event.provider` and `event.dataset`.
> * [log types and mapping](https://opensearch.org/docs/latest/security-analytics/#log-types)

**Comment by mmguero** [comment]
**[<img src="https://avatars.githubusercontent.com/u/13872653?s=17&v=4" width="17" height="17"> @mmguero](https://github.com/mmguero)** [commented](https://github.com/idaholab/Malcolm/issues/162#issuecomment-2392321171) on 2024-10-03: 

> See [this thread](https://forum.opensearch.org/t/null-cannot-be-cast-to-non-null-type-creating-a-detector/21786) for an issue I'm having.

**Comment by mmguero** [comment]
I've narrowed the issue down and have created opensearch-project/security-analytics#1472 as a bug upstream to see if we can get it addressed.

**Comment by mmguero** [comment]
using this command to get a list of unmapped fields, assuming I've created aliases for hte network and host logs:

```bash
for INDEX in malcolm_network malcolm_other; do
    for TOPIC in \
                ad_ldap \
                apache_access \
                azure \
                cloudtrail \
                dns \
                github \
                gworkspace \
                linux \
                m365 \
                netflow \
                network \
                okta \
                others_application \
                others_apt \
                others_cloud \
                others_compliance \
                others_macos \
                others_proxy \
                others_web \
                s3 \
                test_windows \
                vpcflow \
                waf \
                windows; do
        dc \
            exec -u $(id -u) \
            opensearch \
            curl -sS -XGET \
                "http://localhost:9200/_plugins/_security_analytics/mappings/view?index_name=$INDEX&rule_topic=$TOPIC" \
            | jq "del(.unmapped_index_fields) | . += { \"_meta\": {\"index_name\": \"$INDEX\", \"rule_topic\": \"$TOPIC\" } }"
    done
done
```

**Comment by mmguero** [comment]
Tracking some more frustration in [this forum thread](https://forum.opensearch.org/t/some-struggles-creating-mappings-with-api/23373).

**Comment by mmguero** [comment]
[Another bug](https://github.com/opensearch-project/security-analytics/issues/1475) I'm encountering with opensearch security analytics

**Comment by mmguero** [comment]
And [another issue](https://github.com/opensearch-project/security-analytics/issues/1476) I'm finding with security analytics alerts.

**Comment by mmguero** [comment]
And what I think is another bug in trying to create a detector with the [network logs ruleset](https://github.com/opensearch-project/security-analytics/issues/1477).

**Comment by mmguero** [comment]
While we've made a lot of progress on this feature, I think I'm a bit blocked by the following:

* [ ] opensearch-project/security-analytics#1475
* [ ] opensearch-project/security-analytics#1476
* [ ] opensearch-project/security-analytics#1477
* [ ] opensearch-project/security-analytics#1472 (I do have a workaround on this one)

The v25.02.0 will contain the progress we've made but I won't call this "done" until we can resolve those issues and get it working correctly.

**Comment by mmguero** [comment]
[Still jacked up in v3.2.0](https://forum.opensearch.org/t/trouble-with-mappings-detectors-and-alerts/23373/21)
